### PR TITLE
Add scheduler alignment and multi-gpu support for hf_llm example

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Fixes # .

### Description

Upon discussion with CAIA, updated hf_llm example with the following two enhancements:
1. scheduler alignment, mainly done by replacing the epoch increment with HF callback function
2. enable multi-gpu training with resume_from_checkpoint, identified an issue with bnb-related optimizers
3. readme updates reflecting the updates

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/llm_hf/llm_hf_fl_job.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (holgerroth) [inline examples/advanced/llm_hf/llm_hf_fl_job.py]: I think you should keep the original license year if this is not a new file.

File: examples/advanced/llm_hf/README.md [hunk 2]
```diff
@@ -90,35 +91,77 @@ If the correct "global model" is successfully reloaded each round, the training
...
+python3 ./utils/hf_sft_peft_iter.py --output_path ./workspace/dolly_cen_iter 
 ```
 
 The SFT curves are shown below, black for single call, blue for iterative. We can see the "zig-zag" pattern in the iterative training loss curve.
 ![sft](./figs/cen_sft.png)
 
-Similar patterns can be observed from the PEFT curves, purple for single call, green for iterative.
-![peft](./figs/cen_peft.png)
+### Adaptation Step 1.5: iterative training with scheduler alignment and multi-gpu support
+The above iterative training script is a basic version, it works fine if using a constant lr_scheduler. However, if using more complicated lr_scheduler like "cosine_with_restarts", the lr will not be aligned with the iterative training process, which can lead to unexpected results.
```
Comment (holgerroth) [inline examples/advanced/llm_hf/README.md]: "constant learning rate", not "constant lr_scheduler".

**Comment by ZiyueXu77** [comment]
[like]  Ziyue Xu reacted to your message:
________________________________

File: examples/advanced/llm_hf/llm_hf_fl_job.py (hunk 1 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/llm_hf_fl_job.py]: thanks reversed

File: examples/advanced/llm_hf/README.md (hunk 2 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md]: corrected

File: examples/advanced/llm_hf/llm_hf_fl_job.py [hunk 3]
```diff
@@ -29,18 +29,14 @@ def main():
     train_script = "src/hf_sft_peft_fl.py"
     client_ids = args.client_ids
     num_clients = len(client_ids)
+    num_gpus = len(args.gpu.split(","))
```
Comment (holgerroth) [inline examples/advanced/llm_hf/llm_hf_fl_job.py]: This logic will only work for 1 client.

File: examples/advanced/llm_hf/src/hf_sft_peft_fl.py [hunk 4]
```diff
@@ -69,37 +102,61 @@ def main():
...
+    rank, world_size, local_rank = setup_distributed_training()
+
+    # Set up device for DDP
+    device_string = PartialState().process_index
+    device_map = {"": device_string}
+
+    # If output path exists, remove it (only on main process)
+    if local_rank == 0 and os.path.exists(args.output_path):
+        print(f"Output path {args.output_path} exists, removing it.")
+        shutil.rmtree(args.output_path)
```
Comment (holgerroth) [inline examples/advanced/llm_hf/src/hf_sft_peft_fl.py]: shutil is not imported.

File: examples/advanced/llm_hf/README.md
Comment (holgerroth) [inline examples/advanced/llm_hf/README.md:1]: Can you add the command for federated training with multiple clients and multi-gpu?

**Review by holgerroth — CHANGES_REQUESTED** [review]
Nice work. Overall looking good but left some minor comments.

**Review by holgerroth — APPROVED** [review]
